docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+   public PreservedBuffer Preserve();
+}
+
+public struct PreservedBuffer : IDisposable
+{
+   public ReadableBuffer Buffer { get; }
+}
+```
+
+A `ReadCursor` represents a specific coordinate within a `BufferSegment` and a `ReadableBuffer` represents all of the data between 2 `ReadCursors`, Start and End. The `ReadableBuffer` is a point in time snapshot of the data the producer made available. 
+
+After processing the buffer, consumers call `IPipelineReader.Advance(consumed, observed)` indicate how much of the buffer was consumed and how much of the buffer was observed. The distinction is very important as consumed data can be returned to the system for re-use by another component. Marking observed data lets the consumer control over where it should be notified again (to an extent). 
+
+### Threading
+
+When the producer calls `FlushAsync()` on the `WritableBuffer` that will call the consumer on the producer's thread.
+
+### Buffer ownership
+
+In the examples above, data is pushed from the producer to consumer. The consumer has ownership of the buffer between the call to `IPipelineReader.ReadAsync()` and `IPipelineReader.Advance()`. If the consumers needs ownership of the buffer outside of this scope, an explicit call to `ReadableBuffer.Preserve()` must be made before calling Advance. This will transfer ownership of the buffer to the consumer. `PreservedBuffers` must be disposed.
+
+## Problems
+1. When writing higher level protocol parsers, there is a need for efficient low level routines that can grab a byte at a time, and possible several bytes at a time. Further more, to build successful ecosystem of reusable parsing routines, we'd like these types of parsers to be built on the same primitives. Is `ReadableBuffer` that primitive or do we need something lower? 
+2. Multiple buffers. Today pipelines uses a linked list of buffers and provides a single abstraction over them, the `ReadableBuffer`. This multi buffer abstraction means all low level routines need to be implemented against a new multi buffer primitives or duplicated in pipelines.
+3. Structs or classes? Today `WritableBuffer` is tied to the underlying `IPipelineWriter` it was allocated from, this is to enable passing it to other methods without explicitly passing it by reference. This is very limiting as the type is now tied to the underlying writer. `ReadableBuffer` has the same problem. After peeking at bytes in the buffer, you have to make a new slice today because there's no good way to *move* the `ReadableBuffer`. This leads to lots of struct copies and `ReadableBuffer` is a big struct.
+4. Are parsers asynchronous or synchronous? In the model above, parsers are always stateful and synchronous. The code calling into the parser does the awaiting and calls the parser with more data when available. It means that in the future, we'd recommend people write synchronous stateful parsers (JSON.NET etc) and have asynchrony layered on top.
```

Comment:
Synchronous yes. But I am not sure about stateful. I implemented a prototype JSON parser where the caller needs to keep the state and re-pass it when it calls the parser again. If we get "no copy structs" then we can probably make the parsers stateful (as we will be able to make them structs and pass them around).